### PR TITLE
[Done] Refactor pagination and add async.

### PR DESCRIPTION
**Pull request by RoelvandenBerg**

### CODE REVIEW

**Comment by RoelvandenBerg** [comment]
Fixes https://github.com/lizardsystem/lizard-connector/issues/12

File: lizard_connector/connector.py [hunk 1]
```diff
@@ -167,6 +168,67 @@ def header(self):
         return {}
 
 
+class PaginatedRequest(object):
+
+    def __init__(self, endpoint, url):
+        """
```
Comment (GeOdin) [inline lizard_connector/connector.py:175]: Docstring is not compatible with function.

File: lizard_connector/connector.py [hunk 2]
```diff
@@ -179,9 +241,12 @@ def __init__(self, endpoint, base="https://demo.lizard.net", **kwargs):
                             login is used.
             password (str): lizard-api password to log in. Without one no login
                             is used.
-            all_pages (bool): when set to True, on download all pages are
-                              obtained. When set to False only the first
-                              page is obtained on get.
+            paginate (bool): when set to True, paginate results. The connector
```
Comment (GeOdin) [inline lizard_connector/connector.py]: Docstring: endpoint argument is not documented, paginate argument is obsolete.

File: lizard_connector/connector.py [hunk 3]
```diff
@@ -124,27 +143,9 @@ def perform_request(self, url, data=None):
             request_obj = urllib_request.Request(url, headers=self.header)
```
Comment (GeOdin) [inline lizard_connector/connector.py:145]: #141: method="POST"
Not backwards compatible with Python 2.
```
======================================================================
ERROR: test_post (test.test_connector.ConnectorTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/madeleine.vanwinkel/code/lizard-connector/test/test_connector.py", line 75, in test_post
    self.__connector_test(self.connector.post, 'https://test.nl', {'data': 1})
  File "/home/madeleine.vanwinkel/code/lizard-connector/test/test_connector.py", line 67, in __connector_test
    return connector_method(*args, **kwargs)
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 117, in post
    return self.perform_request(url, data)
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 141, in perform_request
    method="POST")
TypeError: __init__() got an unexpected keyword argument 'method'
```

File: lizard_connector/connector.py [hunk 4]
```diff
@@ -203,13 +275,110 @@ def download(self, *querydicts, **queries):
...
+            querydicts (iterable): all key valuepairs from dictionaries are
+                                   used as queries.
+            call_back (function): call back function that is called with the
+                                  downloaded result
+            lock (Lock): a threading lock. This lock is used when executing the
+                         call back function.
+            queries (dict): all keyword arguments are used as queries.
+        """
+        queries.update({"async": "true"})
+        task_url = self.download(*querydicts, **queries).get('url')
```
Comment (GeOdin) [inline lizard_connector/connector.py]:
```
..................Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 372, in async_worker
    task_url = self.download(*querydicts, **queries).get('url')
AttributeError: 'list' object has no attribute 'get'
```

File: lizard_connector/connector.py [hunk 5]
```diff
@@ -124,27 +143,9 @@ def perform_request(self, url, data=None):
             request_obj = urllib_request.Request(url, headers=self.header)
         resp = urlopen(request_obj)
         content = resp.read().decode('UTF-8')
+        # print(content)
```
Comment (GeOdin) [inline lizard_connector/connector.py]: Redundant.

File: lizard_connector/connector.py [hunk 6]
```diff
@@ -27,34 +31,52 @@
...
+    filename = "{}_{}.json".format(file_base, str(int(time.time() * 1000)))
+    with open(filename, 'w') as json_filehandler:
+        json.dump(result, json_filehandler)
+
+
 class Connector(object):
 
-    def __init__(self, max_results=1000, username=None, password=None,
-                 all_pages=True):
+    def __init__(self, max_results=1000, username=None, password=None):
```
Comment (GeOdin) [inline lizard_connector/connector.py]: max_results=1000

This seems unintended.
Pagination seems to work, but if there are more than 1000 assets to download, you get an error.

```
Traceback (most recent call last):
  File "/usr/lib/python3.5/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 228, in __next__
    return self._next_page()
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 203, in _next_page
    self._count = self._endpoint.handle_count(json_)
  File "/home/madeleine.vanwinkel/code/lizard-connector/lizard_connector/connector.py", line 103, in handle_count
    count or 0, self.max_results)
lizard_connector.connector.LizardApiTooManyResults: Too many results: 3259 found, while max 1000 are accepted
```

**Review by GeOdin — DISMISSED** [review]

File: lizard_connector/connector.py (hunk 6 above)
Comment (RoelvandenBerg) [reply lizard_connector/connector.py]: Actually, that is the intended behaviour. You don't want people to request more than a thousand synchronously at once. The error message should be more clear though. Furthermore, you'ld want the max results to differ between async and sync.

File: lizard_connector/connector.py (hunk 5 above)
Comment (RoelvandenBerg) [reply lizard_connector/connector.py]: yup thats ugly

**Comment by jackieleng** [comment]
All checks have failed :(

File: lizard_connector/connector.py (hunk 3 above)
Comment (RoelvandenBerg) [reply lizard_connector/connector.py:145]: Dit is gefixt.

File: lizard_connector/connector.py (hunk 1 above)
Comment (RoelvandenBerg) [reply lizard_connector/connector.py:175]: Dit is gefixt.

**Pull request merged**